cluster-autoscaler/expander/grpcplugin/grpc_client.go

```diff
@@ -0,0 +1,129 @@
+package grpcplugin
+
+import (
+  "context"
+  "log"
+  "time"
+
+  v1 "k8s.io/api/core/v1"
+  "k8s.io/autoscaler/cluster-autoscaler/expander"
+  "k8s.io/autoscaler/cluster-autoscaler/expander/grpcplugin/protos"
+  "k8s.io/klog/v2"
+  schedulerframework "k8s.io/kubernetes/pkg/scheduler/framework"
+
+  "google.golang.org/grpc"
+  "google.golang.org/grpc/credentials"
+)
+
+type grpcclientstrategy struct {
+  grpcClient protos.ExpanderClient
+}
+
+type GRPCOpts struct {
+  GRPCExpanderCert string
+  GRPCExpanderUrl string
+}
+
+// NewFilter returns an expansion filter that creates a gRPC client, and calls out to a gRPC server
+func NewFilter(opts GRPCOpts) expander.Filter {
+  client := createGRPCClient(opts)
+  if client == nil {
+    return &grpcclientstrategy{grpcClient: nil}
+  }
+  return &grpcclientstrategy{grpcClient: client}
+}
+
+func createGRPCClient(opts GRPCOpts) protos.ExpanderClient {
+  var dialOpt grpc.DialOption
+
+  // if no Cert file specified, use insecure
+  if opts.GRPCExpanderCert == "" {
+    dialOpt = grpc.WithInsecure()
+  } else {
+    creds, err := credentials.NewClientTLSFromFile(opts.GRPCExpanderCert, "")
+    if err != nil {
+      log.Fatalf("Failed to create TLS credentials %v", err)
+      return nil
+    }
+    dialOpt = grpc.WithTransportCredentials(creds)
+  }
+  klog.V(1).Info("Dialing ", opts.GRPCExpanderUrl, " dialopt: ", dialOpt)
+  conn, err := grpc.Dial(opts.GRPCExpanderUrl, dialOpt)
+  if err != nil {
+    log.Fatalf("fail to dial server: %v", err)
+    return nil
+  }
+  return protos.NewExpanderClient(conn)
+}
+
+func (g *grpcclientstrategy) BestOptions(expansionOptions []expander.Option, nodeInfo map[string]*schedulerframework.NodeInfo) []expander.Option {
+  if g.grpcClient == nil {
+    log.Fatalf("Incorrect gRPC client config, filtering no options")
+    return expansionOptions
+  }
+
+  // Transform inputs to gRPC inputs
+  nodeGroupIDOptionMap := make(map[string]expander.Option)
+  grpcOptionsSlice := []*protos.Option{}
+  populateOptionsForGRPC(expansionOptions, nodeGroupIDOptionMap, &grpcOptionsSlice)
+  grpcNodeInfoMap := make(map[string]*v1.Node)
+  populateNodeInfoForGRPC(nodeInfo, grpcNodeInfoMap)
+
+  // call gRPC server to get BestOption
+  klog.V(1).Info("GPRC call of best options to server with ", len(nodeGroupIDOptionMap), " options")
```

Comment:
Not 100% sure if "1" is the right log level here, might be worth double-checking here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md